### PR TITLE
[MacOS]Added clear button in date and time

### PR DESCRIPTION
**Pull request by mukuagar**
## Description
Create a clear button

At the very least please describe the issue you are addressing, and what your fix entails. 

## Sample Card
image
Date present X mark shown else not

image
Same for time

## How Verified
Added new tests

### CODE REVIEW

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift [hunk 1]
```diff
@@ -178,6 +201,7 @@ class ACRDateField: NSView, InputHandlingViewProtocol {
         selectedDate = datePicker.dateValue
         datePickerCalendar.dateValue = datePicker.dateValue
         datePickerTextfield.dateValue = datePicker.dateValue
+        clearButton.isHidden = false
     }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:205]: Guess this is not required

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift [hunk 2]
```diff
@@ -122,6 +138,13 @@ class ACRDateField: NSView, InputHandlingViewProtocol {
         addTrackingArea(trackingArea)
     }
     
+    @objc private func handleClearAction() {
+        textField.stringValue = ""
+        dateValue = ""
+        selectedDate = .none
+        clearButton.isHidden = true
+    }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:146]: Why are we setting `isHidden` property in multiple places. It should have only one single source of truth.

```suggestion
        dateValue = nil
    }
```
This should be enough here.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift [hunk 3]
```diff
@@ -60,6 +70,9 @@ class ACRDateField: NSView, InputHandlingViewProtocol {
             if let dateValue = dateValue, let date = dateFormatter.date(from: dateValue) {
                 textField.stringValue = dateFormatterOut.string(from: date)
                 selectedDate = date
+                clearButton.isHidden = false
+            } else {
+                clearButton.isHidden = true
             }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:76]:
```suggestion
            } else {
                textField.stringValue = ""
                clearButton.isHidden = true
            }
```

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift [hunk 4]
```diff
@@ -15,7 +15,7 @@ class InputDateRendererTest: XCTestCase {
     }
 
     func testRendererSetsValue() {
-        let val: String = "23/11/2000"
+        let val: String = "2000-10-12"
         inputDate = .make(value: val)
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:19]: Why are these changed?
Was the change causing any regression?

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift [hunk 5]
```diff
@@ -31,20 +31,55 @@ class InputDateRendererTest: XCTestCase {
...
+        XCTAssertTrue(inputDateField.clearButton.isHidden)
+    }
+    
+    func testClearButtonHiddenWithPlaceholder() {
+        let placeholderString: String = "Sample Placeholder"
+        inputDate = .make(placeholder: placeholderString)
+
+        let inputDateField = renderDateInput()
+        inputDateField.clearButton.performClick()
+        XCTAssertTrue(inputDateField.clearButton.isHidden)
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:65]: Why do we need to call `clearButton.performClick()` here??

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift [hunk 6]
```diff
@@ -31,20 +31,55 @@ class InputDateRendererTest: XCTestCase {
...
+    }
+    
+    func testClearButtonHiddenWithNoText() {
+        inputDate = .make()
+        
+        let inputDateField = renderDateInput()
+        inputDateField.clearButton.performClick()
+        XCTAssertTrue(inputDateField.clearButton.isHidden)
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:73]: Same here. Why is `performClick()` needed??
It shouldn't be called as button will be already hidden. Just checking whether it's hidden is enough.

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift [hunk 7]
```diff
@@ -55,3 +90,10 @@ class InputDateRendererTest: XCTestCase {
     }
 }
 
+private extension NSButton {
+    func performClick() {
+        if let target = target, let action = action {
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:95]:
```suggestion
    func performClick() {
        guard !isHidden else { return }
        if let target = target, let action = action {
```
Maybe add this here. I see `performClick()` should not work if the button is hidden.

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputTimeRendererTests.swift [hunk 8]
```diff
@@ -45,6 +45,42 @@ class InputTimeRendererTest: XCTestCase {
...
+        inputTimeField.clearButton.performClick()
+        XCTAssertTrue(inputTimeField.clearButton.isHidden)
+    }
+
+    func testClearButtonHiddenWithNoText() {
+        inputTime = .make()
+
+        let inputTimeField = renderTimeInput()
+        inputTimeField.clearButton.performClick()
+        XCTAssertTrue(inputTimeField.clearButton.isHidden)
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputTimeRendererTests.swift:73]:
```suggestion

        let inputTimeField = renderTimeInput()
        XCTAssertTrue(inputTimeField.clearButton.isHidden)
    }

    func testClearButtonHiddenWithNoText() {
        inputTime = .make()

        let inputTimeField = renderTimeInput()
        XCTAssertTrue(inputTimeField.clearButton.isHidden)
```

Use `performClick()` only in places where it's required to simulate a click.
These cases doesn't make sense to have click events.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift (hunk 3 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:76]: The string value is already blank, that is why I didn't initialise it

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift (hunk 1 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:205]: Akaash without this
The clear button won't reappear once date is picked again without this

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift (hunk 2 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRDateField.swift:146]: we need to also set the textfield.stringvalue and clearButton.isHidden here
This handles the general case and the other clearButton.isHidden = true is for the specific case when there is an initial value already stored

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift (hunk 4 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:19]: Yeah, the formatter expects the date in yyyy-mm-dd format
dd/mm/yyyy was the old format which should not be supported. It was fixed in the code but never changed here as the tests were still passing I guess

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift (hunk 5 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:65]: Yeah true
Even without it it will work
I'll change it in this and inputtimetests

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift (hunk 7 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:95]: Oh thanks Akaash

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputTimeRendererTests.swift (hunk 8 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputTimeRendererTests.swift:73]: Sure will do thanks!

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift (hunk 6 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Renderers/InputDateRendererTests.swift:73]: True
Will change

**Comment by blue-stone99** [comment]
Changes are 🔥
